firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/TransformMutation.java

```diff
@@ -123,23 +121,28 @@ public MaybeDocument applyToLocalView(
         getKey(), doc.getVersion(), newData, Document.DocumentState.LOCAL_MUTATIONS);
   }
 
+  @Nullable
   @Override
-  public FieldMask getFieldMask() {
-    Set<FieldPath> fieldMask = new HashSet<>();
-    for (FieldTransform transform : fieldTransforms) {
-      fieldMask.add(transform.getFieldPath());
-    }
-    return FieldMask.fromSet(fieldMask);
-  }
+  public ObjectValue extractBaseValue(@Nullable MaybeDocument maybeDoc) {
+    ObjectValue baseValue = null;
 
-  @Override
-  public boolean isIdempotent() {
     for (FieldTransform transform : fieldTransforms) {
-      if (!transform.isIdempotent()) {
-        return false;
+      FieldValue previousValue = null;
+      if (maybeDoc instanceof Document) {
+        previousValue = ((Document) maybeDoc).getField(transform.getFieldPath());
+      }
+
+      FieldValue value = transform.getOperation().computeBaseValue(previousValue);
```

Comment:
We have a lot of "value" stuff going on here: extractBaseValue(), baseValue, previousValue, value, computeBaseValue() and it's a bit misleading and hard to keep track of. I think we're probably overloading "baseValue" a bit. You tried to differentiate via verb (compute vs extract), but it might be better to change the noun...

So I'm wondering if we could rename "baseValue" in "extractBaseValue()" to make it a bit clearer that it's a "composite" value to be used as a patch mutation...  extractBaseMutationValue()?  extractBaseValues()?

Idunno. WDYT?

Reply:
baseValue => baseObject, previousValue => existingValue, value => coercedValue (?)

Reply:
Maaaaaaybe.  The current code includes all of the following in different places: existingValue, coercedValue, previousValue, currentValue, baseObject, baseValue, maybeDoc, baseDoc  ...  which I'm pretty certain is more terms than we have concepts.

Can you see if we can get rid of previousValue in applyToLocalView()?  And then I'll take another look at the naming.